File: common/src/main/java/net/impleri/playerskills/integrations/rei/PlayerSkillsReiPlugin.java
package net.impleri.playerskills.integrations.rei;

import me.shedaniel.rei.api.client.plugins.REIClientPlugin;
import me.shedaniel.rei.api.client.registry.display.DisplayRegistry;
import net.impleri.playerskills.PlayerSkills;
import net.impleri.playerskills.client.PlayerSkillsClient;
import net.impleri.playerskills.client.restrictions.RecipeRestrictionOpsClient;
import net.impleri.playerskills.restrictions.RestrictionRegistry;
import net.impleri.playerskills.utils.PlayerSkillsLogger;
import org.jetbrains.annotations.NotNull;

/**
 * This is necessary because the REIPlugin::compareTo function uses a raw type and scala does not like it.
 */
public class PlayerSkillsReiPlugin implements REIClientPlugin {
  private final @NotNull RestrictionRegistry restrictions;
  private final @NotNull RecipeRestrictionOpsClient recipeOps;

  public PlayerSkillsReiPlugin() {
    restrictions = PlayerSkills.STATE().RESTRICTIONS();
    recipeOps = PlayerSkillsClient.STATE().RECIPE_RESTRICTIONS();
  }

  @Override
  public void registerDisplays(DisplayRegistry registry) {
    registry.registerVisibilityPredicate(RecipeRestrictionsPredicate.apply(registry, restrictions, recipeOps, PlayerSkillsLogger.RECIPES()));
  }
}

File: common/src/main/kotlin/net/impleri/playerskills/integrations/rei/PlayerSkillsReiPlugin.kt
package net.impleri.playerskills.integrations.rei

import me.shedaniel.rei.api.client.entry.filtering.base.BasicFilteringRule
import me.shedaniel.rei.api.client.plugins.REIClientPlugin
import me.shedaniel.rei.api.client.registry.display.DisplayRegistry
import net.impleri.playerskills.events.ClientSkillsUpdatedEvent

open class PlayerSkillsReiPlugin : REIClientPlugin {
  init {
    // Subscribe to client skill updates to refresh basic filtering rules
    ClientSkillsUpdatedEvent.EVENT.register {
      PlayerSkillsFiltering.updateHidden(it)
    }
  }

  override fun registerDisplays(registry: DisplayRegistry) {
    registry.registerVisibilityPredicate(SkillsDisplayVisibility())
  }

  override fun registerBasicEntryFiltering(rule: BasicFilteringRule<*>) {
    PlayerSkillsFiltering.register(rule)
  }
}
